bpf/bpf_host.c

```diff
@@ -760,34 +976,79 @@ int from_host(struct __ctx_buff *ctx)
 __section("to-netdev")
 int to_netdev(struct __ctx_buff *ctx __maybe_unused)
 {
-	/* Cannot compile the section out entriely, test/bpf/verifier-test.sh
-	 * workaround.
-	 */
+	__u32 __maybe_unused remoteID = WORLD_ID, srcID = 0;
+	__u16 __maybe_unused proto = 0;
 	int ret = CTX_ACT_OK;
+
 #if defined(ENABLE_NODEPORT) && \
 	(!defined(ENABLE_DSR) || \
 	 (defined(ENABLE_DSR) && defined(ENABLE_DSR_HYBRID)))
-	if ((ctx->mark & MARK_MAGIC_SNAT_DONE) == MARK_MAGIC_SNAT_DONE)
-		return CTX_ACT_OK;
-	ret = nodeport_nat_fwd(ctx, false);
-	if (IS_ERR(ret))
-		return send_drop_notify_error(ctx, 0, ret, CTX_ACT_DROP, METRIC_EGRESS);
+	if ((ctx->mark & MARK_MAGIC_SNAT_DONE) != MARK_MAGIC_SNAT_DONE) {
+		ret = nodeport_nat_fwd(ctx, false);
+		if (IS_ERR(ret))
+			return send_drop_notify_error(ctx, 0, ret,
+						      CTX_ACT_DROP,
+						      METRIC_EGRESS);
+	}
 #endif
+
+#ifdef ENABLE_HOST_FIREWALL
+	if (!proto && !validate_ethertype(ctx, &proto)) {
+		ret = DROP_UNSUPPORTED_L2;
+		goto out;
+	}
+
+	policy_clear_mark(ctx);
+
+	switch (proto) {
+# if defined ENABLE_ARP_PASSTHROUGH || defined ENABLE_ARP_RESPONDER
+	case bpf_htons(ETH_P_ARP):
+		ret = CTX_ACT_OK;
+		break;
+# endif
+# ifdef ENABLE_IPV6
+	case bpf_htons(ETH_P_IPV6):
+		// TODO ipv6_host_policy_ingress.
+		ret = CTX_ACT_OK;
+		break;
+# endif
+# ifdef ENABLE_IPV4
+	case bpf_htons(ETH_P_IP):
+		// to-netdev is attached to the egress path of the native
+		// device.
+		srcID = ipcache_lookup_srcid4(ctx);
+		ret = ipv4_host_policy_egress(ctx, srcID, &remoteID);
```

Comment:
:thinking: We would need some sort of mark to tell the host firewall to let those packets through. Don't we already have similar mechanisms to carry info. through non-BPF NATing?

Reply:
Right, so for reference the way we do this today is basically we have an OUTPUT chain rule that explicitly marks the traffic as "from the host". OUTPUT chain implies that the socket is in the local network namespace so the process *must* be local. We can then treat all traffic without this mark as "remote" in some sense. If traffic from the outside world arrives to the node and is SNAT'd, then it will not hit this rule. When we handle this traffic, even if we look up the IPCache and it maps the IP to host identity, we will explicitly ignore the ipcache and treat the traffic as from the world instead.

One tricky part here is that we are not in control of the component which installs the SNAT iptables rules here, so we can't just modify those rules to provide this context.

Reply:
Ah, right, makes sense. Are we able to carry a mark through kube-proxy's SNATing though?

Maybe that's something to discuss at the next sig-datapath meeting :thinking:

Reply:
Passing a mark through iptables rules established by other agents like kube-proxy can potentially be done but we don't have a lot of control here. A newer release of the other agent could always modify the way they deploy iptables in a way that breaks our use of the mark. This is part of the motivation for the way we do the marking currently.

But yeah this is a bigger topic than this PR and is one of the last remaining items for eliminating Cilium's reliance on iptables, good candidate for sig-datapth as it could get fairly deep.

Reply:
(I might add; the mark bits we *do* use are selected because they don't overlap with the semantics of mark bits used by kube-proxy, so yes we can and do carry marks through kube-proxy's SNAT rules but there's definitely details to work through there)